In [39]:
!pip install selenium

Сначала нами для парсинга был применен пакет BeautifulSoup, однако он выдавал ошибку 403, что говорит о том, что мы не сможем собрать данные. Так как Циан является одним из наиболее популярных сайтов недвижимости в России, в последние годы он начал активно заниматься усилением своей защиты, поэтому нам пришлось использовать пакет selenium. Данный пакет мы установили с помощью selenium.webdriver.

In [40]:
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.common.keys import Keys
import pandas as pd
import time
from bs4 import BeautifulSoup
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC

In [41]:
from selenium.webdriver import Chrome
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.support.ui import WebDriverWait as wait
from selenium.webdriver.support import expected_conditions as EC
from selenium.common.exceptions import NoSuchElementException
from bs4 import BeautifulSoup
from time import sleep
from tqdm import tqdm

In [58]:
import selenium
selenium.__version__

'4.8.3'

Для проекта нами были выбраны квартиры в новостройках в Москве с количеством комнат 1-6+

Так как сайт является динамическим и объявления постоянно обновляются отметим дату выгрузки 6.04.2023

Далее ввыгрузим ссылки с объвлениями, по установленной фильтрации. Также стоит отметить, что мы рассматриваем 55 страниц с объявлениями, так как нами при проверке ссылок на уникальность было выявлено, что 56 страницы не существует, и сайт автоматически перебрасывает на 1 страницу с объявлениями, поэтому мы ограничили наши объявления 55 страницами.

Мы выгрузили все ссылки, чтобы идейно посмотреть все ли делаем верно и работает ли код. Однако, чтобы мы смогли зайти в каждую из ссылок, нам необходимо создать список, чтобы последовательно заходить в каждую из ссылок и выгружать данные с них.

In [7]:
browser = webdriver.Chrome(r'C:/Users/sashka/Documents/Универ/Андан')
for p in range(1,55):
    url = f'https://www.cian.ru/cat.php?deal_type=sale&engine_version=2&object_type%5B0%5D=2&offer_type=flat&p={p}&region=1&room1=1&room2=1&room3=1&room4=1&room5=1&room6=1'
    browser.get(url)

    for a in  browser.find_elements(By.CLASS_NAME, "_93444fe79c--media--9P6wN"):
        print(a.get_attribute('href'))

/var/folders/9r/50t5jc150bv76h_18l7x10k40000gn/T/ipykernel_99943/2513475264.py:1: DeprecationWarning: executable_path has been deprecated, please pass in a Service object
  browser = webdriver.Chrome(r'C:/Users/sashka/Documents/Универ/Андан')


https://www.cian.ru/sale/flat/283969757/
https://www.cian.ru/sale/flat/284037342/
https://www.cian.ru/sale/flat/266826884/
https://www.cian.ru/sale/flat/284807290/
https://www.cian.ru/sale/flat/264958724/
https://www.cian.ru/sale/flat/282960185/
https://www.cian.ru/sale/flat/283456071/
https://www.cian.ru/sale/flat/281325484/
https://www.cian.ru/sale/flat/274260488/
https://www.cian.ru/sale/flat/285461914/
https://www.cian.ru/sale/flat/272591258/
https://www.cian.ru/sale/flat/285682617/
https://www.cian.ru/sale/flat/274969991/
https://www.cian.ru/sale/flat/285606751/
https://www.cian.ru/sale/flat/282699510/
https://www.cian.ru/sale/flat/279224278/
https://www.cian.ru/sale/flat/268828538/
https://www.cian.ru/sale/flat/280755821/
https://www.cian.ru/sale/flat/277738124/
https://www.cian.ru/sale/flat/285442011/
https://www.cian.ru/sale/flat/262519487/
https://www.cian.ru/sale/flat/284404564/
https://www.cian.ru/sale/flat/281896678/
https://www.cian.ru/sale/flat/284824229/
https://www.cian

In [8]:
browser = webdriver.Chrome(r'C:/Users/sashka/Documents/Универ/Андан')
links = []
for p in range(1,55):
    url = f'https://www.cian.ru/cat.php?deal_type=sale&engine_version=2&object_type%5B0%5D=2&offer_type=flat&p={p}&region=1&room1=1&room2=1&room3=1&room4=1&room5=1&room6=1'
    browser.get(url)
#browser.find_element("_93444fe79c--color_primary_100--mNATk", class)
    for a in  browser.find_elements(By.CLASS_NAME, "_93444fe79c--media--9P6wN"):
        links.append(a.get_attribute('href'))
print(links)

/var/folders/9r/50t5jc150bv76h_18l7x10k40000gn/T/ipykernel_99943/3016437563.py:1: DeprecationWarning: executable_path has been deprecated, please pass in a Service object
  browser = webdriver.Chrome(r'C:/Users/sashka/Documents/Универ/Андан')


['https://www.cian.ru/sale/flat/283969757/', 'https://www.cian.ru/sale/flat/284037342/', 'https://www.cian.ru/sale/flat/266826884/', 'https://www.cian.ru/sale/flat/284807290/', 'https://www.cian.ru/sale/flat/264958724/', 'https://www.cian.ru/sale/flat/282960185/', 'https://www.cian.ru/sale/flat/283456071/', 'https://www.cian.ru/sale/flat/281325484/', 'https://www.cian.ru/sale/flat/274260488/', 'https://www.cian.ru/sale/flat/285461914/', 'https://www.cian.ru/sale/flat/272591258/', 'https://www.cian.ru/sale/flat/285682617/', 'https://www.cian.ru/sale/flat/274969991/', 'https://www.cian.ru/sale/flat/285606751/', 'https://www.cian.ru/sale/flat/282699510/', 'https://www.cian.ru/sale/flat/279224278/', 'https://www.cian.ru/sale/flat/268828538/', 'https://www.cian.ru/sale/flat/280755821/', 'https://www.cian.ru/sale/flat/277738124/', 'https://www.cian.ru/sale/flat/285442011/', 'https://www.cian.ru/sale/flat/262519487/', 'https://www.cian.ru/sale/flat/284404564/', 'https://www.cian.ru/sale/flat/

In [55]:
with open('cian_final.txt') as f:
    lines = f.readlines()

In [58]:
len(lines)

1509

Мы выгрузили 1509 объявлений, однако len(lines) считает объявления не с 0, а с 1, поэтому берем на одно меньше.

Импортируем функцию tqdm из модуля notebook в библиотеке tqdm. Эта функция используется для отображения прогресса выполнения цикла в блокноте Jupyter Notebook.

In [59]:
len(lines)
from tqdm.notebook import tqdm

***Код для создания файла csv***

Теперь выгрузим ссылки в csv файл, чтобы мы могли выгрузить наши данные для дальнейшей их обработки.

In [44]:
with open('cian_final.txt', 'w') as file:
    for i in range(len(links)):
        file.write(f'{links[i]}\n')
file.close()

***Выгрузка данных***


Теперь выгрузим все характеристики и их значения в csv файл, проходя по ссылки каждого объявления и вытаскивая, нужные нам данные. По идее в ранге должны стоять данные (0, 1508), но комп решил выключиться, и выгрузка данных остановилась на 1505 объявлении, поэтому в данном коде мы берем range(1506,1508) (до этого range был 1508).

После изучения структуры объявлений на сайте Циана, нам стало понятно, какие данные / блоки данных, нужно выгружать. 

Также мы столкнулись с трудностью, что не во всех объявлениях есть все характеристики, поэтому код ломался на 85 объявлении. Для того, чтобы он прошел и выгрузил все данные вне зависимости есть ли та или иная характеристика, мы используем команду except.

In [67]:
import csv
browser = webdriver.Chrome(r'C:/Users/sashka/Documents/Универ/Андан')
#выгрузка и запись на 10 первых
with open('last_try.csv', 'a', newline='\n', encoding="utf-8") as data:
    writer = csv.writer (data)
    for link in tqdm(range(1506,1508)):
        para_to=[]
        a=lines[link][:-1]
    #создаём первуую колонку (номер объявления)
        para_to.append(a.split('/')[-2])
        browser.get(a)
        sleep(2)
    #зашли на страницу объявления
        try:
            para_to.append((browser.find_element(By.CLASS_NAME, "a10a3f92e9--title--vlZwT")).text[0])
        except:
            para_to.append('none')
    #количество комнат
        try:
            para_to.append((browser.find_element(By.CLASS_NAME, "a10a3f92e9--price--Pg6fn")).text)
        except:
            para_to.append('none')
    #цена
        try:
            para_to.append((browser.find_element(By.CLASS_NAME, "a10a3f92e9--underground_link--VnUVj")).text)
        except:
            para_to.append('none')
        try:
            para_to.append((browser.find_element(By.CLASS_NAME, "a10a3f92e9--underground_time--YvrcI")).text)
        except:
            para_to.append('none')
    #метро+время до метро, берём первое, потому что циан сам сортирует время по возрастанию (нужно добавить разделение на машине/пешком)
        try:
            c=(browser.find_element(By.CLASS_NAME, "a10a3f92e9--container--tqDAE")).text
            para_to.append(c.split('\n'))
        except:
            para_to.append('none')
            
        try:
            d=(browser.find_element(By.CLASS_NAME, "a10a3f92e9--container--rGqFe")).text
            para_to.append(d.split('\n'))
        except:
            para_to.append('none')
            
        writer.writerow(para_to)
        
data.close()

/var/folders/9r/50t5jc150bv76h_18l7x10k40000gn/T/ipykernel_1004/776354650.py:2: DeprecationWarning: executable_path has been deprecated, please pass in a Service object
  browser = webdriver.Chrome(r'C:/Users/sashka/Documents/Универ/Андан')


  0%|          | 0/2 [00:00<?, ?it/s]

In [96]:
slovar = dict()
print(slovar)

{}


На данном шаге мы создадим пустой dataframe.csv и выгрузим туда уже все наши объявления с характеристиками.

На данном этапе мы заканчиваем работу с формированием csv файла и переходим к работе над ним.

In [113]:
with open('dataframe.csv') as f:
    lines = f.readlines()

In [104]:
slovar = []
for line in lines:
    case = {'Номер объявления': 283969757, 'Количество комнат': 1, 'Цена': 16042265 }
    slovar.append(case)

In [105]:
# Python program to read CSV file line by line
# import necessary packages

import csv
  
# Open file 
with open('dataframe.csv') as f:
      
    # Create reader object by passing the file 
    # object to reader method
    reader_obj = csv.reader(f)
      
    # Iterate over each row in the csv 
    # file using reader object
    for row in reader_obj:
        print(row)

['283969757', '1', '16 042 265 ₽', 'Савёловская', '13 мин.', "['Общая площадь', '32,2 м²', 'Этаж', '4 из 22', 'Год сдачи', '2025', 'Дом', 'Не сдан', 'Отделка', 'Без отделки']", "['О квартире', 'Тип жилья', 'Новостройка', 'Общая площадь', '32,2', 'Высота потолков', '3 м', 'Отделка', 'Без отделки', 'О доме', 'Подписаться на дом', 'Количество лифтов', '1 пассажирский', 'Тип дома', 'Монолитный', 'Парковка', 'Подземная']"]
['284037342', '1', '16 297 308 ₽', 'Технопарк', '17 мин.', "['Общая площадь', '43,5 м²', 'Этаж', '2 из 17', 'Год сдачи', '2024', 'Дом', 'Не сдан', 'Отделка', 'Без отделки']", "['О квартире', 'Тип жилья', 'Новостройка', 'Общая площадь', '43,5', 'Высота потолков', '3,1 м', 'Отделка', 'Без отделки', 'О доме', 'Подписаться на дом', 'Тип дома', 'Монолитно-кирпичный', 'Парковка', 'Подземная']"]
['266826884', '1', '17 086 799 ₽', 'Дмитровская', '8 мин.', "['Общая площадь', '32,1 м²', 'Жилая площадь', '11,2 м²', 'Площадь кухни', '13,4 м²', 'Этаж', '11 из 36', 'Год сдачи', '2025',